\documentclass{bioinfo}
\copyrightyear{2015} \pubyear{2015}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Manuscript Category}

\begin{document}
\firstpage{1}

\subtitle{Data and text mining}

\title[iSEE]{iSEE: Interactive SummarizedExperiment Explorer}
\author[Rue-Albrecht \textit{et~al}.]{Kevin Rue-Albrecht\,$^{\text{\sfb 1}}$,
    Federico Marini\,$^{\text{\sfb 2}}$,
    Charlotte Soneson\,$^{\text{\sfb 3}}$
    and Aaron T. L. Lun\,$^{\text{\sfb 4,}*}$} 
\address{$^{\text{\sf 1}}$Kennedy Institute of Rheumatology, University of Oxford, Oxford OX3 7FY, United Kingdom, \\
$^{\text{\sf 2}}$Institute for Medical Biostatistics, Epidemiology and Informatics, \\
$^{\text{\sf 3}}$University of Zurich and SIB Swiss Institute of Bioinformatics, and \\
$^{\text{\sf 4}}$Cancer Research UK Cambridge Institute, University of Cambridge, Cambridge CB2 0RE, United Kingdom}

\corresp{$^\ast$To whom correspondence should be addressed.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\abstract{\textbf{Motivation:} Text Text Text Text Text Text Text Text Text Text Text Text Text
Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text
Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text
Text Text Text Text Text Text
Text Text Text Text Text.\\
\textbf{Results:} Text  Text Text Text Text Text Text Text Text Text  Text Text Text Text Text
Text Text Text Text Text Text Text Text Text Text Text Text Text  Text Text Text Text Text Text\\
\textbf{Availability:} Text  Text Text Text Text Text Text Text Text Text  Text Text Text Text
Text Text Text Text Text Text Text Text Text Text Text Text Text Text  Text\\
\textbf{Contact:} \href{aaron.lun@cruk.cam.ac.uk}{aaron.lun@cruk.cam.ac.uk}\\
\textbf{Supplementary information:} Supplementary data are available at \textit{Bioinformatics}
online.}

\maketitle

\section{Introduction}
Interactive data exploration is critical to the analysis and comprehension of the data generated by high-throughput biological assays such as transcriptomics. 
Exploration drives the formation of novel data-driven hypotheses prior to a more rigorous statistical analysis, and enables diagnosis of potential problems such as batch effects, contaminants and low-quality samples. 
To this end, rapid visualization of the data in an intuitive and interactive interface is crucial, allowing researchers to examine the data from different perspectives across samples (e.g., experimental replicates, individual patients, single cells) and/or features (e.g., genes, transcripts, proteins, genomic regions).

Most existing tools for interactive visualization of biological data are designed for specific assays or analyses, e.g., pRoloc for proteomics data \citep{gatto2014mass}, shinyMethyl for methylation data \citep{fortin2014shinymethyl}, HTSvis for high-throughput screens \citep{scheeder2017htsvis}.
This makes it difficult to recycle the same visualization framework for new technologies or experimental designs where different aspects of the data are of interest. 
Moreover, standalone tools such as XXX do not easily integrate into established analysis frameworks such as those based on the R statistical programming language (CITE R Core Team). 
This complicates any coordinated use of these tools with a reproducible, transparent and rigorous statistical analysis. 

\section{Description of the iSEE interface}
Here, we present the iSEE software package for interactive exploratory data analysis. 
iSEE is implemented in R using the interactive shiny framework and provides a number of features such as a customizable multi-panel layout, dynamic linking for point selection between plots, code reproducibility and intelligent downsampling. 
iSEE exploits the well-established SummarizedExperiment data architecture in the open-source Bioconductor project \citep{gentleman2004bioconductor}.
This enables iSEE to accommodate a wide range of `-omics datasets that are routinely processed by Bioconductor packages, ranging from conventional RNA sequencing to single-cell transcriptomics to mass cytometry.

\subsection{Integration with Bioconductor packages} 
The iSEE interface is initialized with a single function call accepting a SummarizedExperiment object as input \citep{huber2015orchestrating}.
This object stores one or more matrices of experimental observations as ``assays'', where columns and rows represent samples and features, respectively. 
Per-feature or per-sample variables can also be easily added to the object in the ``rowData'' and ``colData'', respectively; this includes experimental metadata as well as any analysis results. 
The flexibility of the SummarizedExperiment class is the driving factor behind its wide deployment throughout the Bioconductor ecosystem, with uses in RNA sequencing \citep{love2014moderated}, ChIP-seq \citep{lun2016csaw}, methylation \citep{aryee2014minfi}, Hi-C \citep{lun2016infrastructure} and mass cytometry (PMID: 28504682, lukas). 
Developers can easily derive subclasses for specialized use cases, such as the SingleCellExperiment class that enables easy storage of sample-wise reduced dimensionality embeddings in single-cell `-omics data analyses. 
By accepting a SummarizedExperiment object, iSEE immediately offers interactive visualization to complement existing analysis pipelines for a variety of data modalities. 
In particular, we leverage off the state-of-the-art statistical and computational methods already available in the R/Bioconductor framework, allowing us to avoid re-implementing them in iSEE (as is commonly done in other standalone visualization tools).

\subsection{Visualizing multiple data aspects simultaneously}
iSEE uses a customisable multi-panel layout that simultaneously displays one or more panels of different types, where each panel type visualizes a different aspect of the data. 
This includes metadata for each feature or sample, experimental observations across samples for each feature, and reduced dimensionality embeddings for each sample. 
For most panels, each sample or feature is represented as a point in a plot with the specified variables on the x/y-axes.
Depending on whether these variables are categorical or numeric, panels will automatically switch between scatter plots, violin plots or ``rectangle plots'' (where each combination of categorical levels is represented by a rectangle with area proportional to the number of samples/features with that combination). 
Each point can also be coloured according to data aspects such as sample-level metadata or feature-level observations. 
Other panel types include heatmaps, to visualize experimental observations for multiple features; and row statistics tables, to facilitate searching across features and their metadata.
All panels can be added, removed, reordered or resized.

\subsection{Dynamic linking across panels}
A key feature of iSEE is the ability to dynamically transmit information between panels. 
Users can define arbitrary links between a ``transmitting'' and ``receiving'' plot, whereby selection of points on the transmitting plot will highlight the corresponding points in the receiving plot (or optionally, hide all unselected points). 
This facilitates exploration of the relationships between different aspects of the data -- for example, users can easily determine co-expression patterns of genes in a particular region of a reduced dimensionality embedding, by transmitting information from the reduced dimension plot to a feature expression plot. 
The linking paradigm extends to multiple plots whereby a plot can transmit to multiple receivers, or a receiving plot can itself transmit to another plot.
This allows users to mimic the arbitrarily complex gating strategies often found in analyses of flow cytometry data \citep{finak2014opencyto}, with the advantage of being extensible to any (meta)data present in a SummarizedExperiment object. 
Information can similarly be transferred between tables and plots, e.g., selecting a row determines how points are coloured in receiving plots.

\subsection{Customizing the user experience}
iSEE also offers a host of other customizable features to improve data exploration. 
Users can supply their own colour mappings to iSEE, to control how colour scales are chosen for each individual  metadata variable and assay type consistently across all panels in the application. 
Row statistics tables can be augmented with dynamic annotation based on the selected row, providing links to online resources such as Ensembl or Entrez annotation. 
For large datasets, points can be downsampled in a density-dependent manner to accelerate rendering of the plots, improving the responsiveness of the interface without compromising the fidelity of the visualization. 
After analyses are completed, users can deploy iSEE on a server with a bespoke step-by-step ``tour'' of their dataset, guiding the audience through an examination of the salient features in the data.

\subsection{Reproducing the plotting code}
iSEE memorizes the exact R code that was used to generate every plot. 
This allows users to easily reproduce the results of any exploratory analysis, simply by copying the code reported by iSEE into their own analysis scripts. 
All plots are generated using the ggplot2 framework and can be easily customized. 

\section{Conclusion}
iSEE provides a general interactive interface for visual exploration of a diverse range of `-omics datasets. 
We provide a number of case studies at (REPOSITORY NAME) to demonstrate how iSEE can be used to explore the TCGA RNA-seq data set (CITE someone?), the 68,000 PBMC single-cell RNA sequencing dataset from 10X Genomics (CITE Zheng) and a published mass cytometry dataset consisting of over 170,000 cells (CITE?). 
Interested users can also easily apply iSEE on their own data by downloading the package from Bioconductor (https://bioconductor.org/packages/iSEE).

\section*{Acknowledgements}
We thank the organizers of the European Bioconductor Meeting 2017, where the idea for this package was first conceived.
We also thank members of the Bioconductor community for their helpful suggestions. 
\vspace*{-12pt}

\section*{Funding}
ATLL was supported by core funding from Cancer Research UK (award no. 17197 to Dr.\ John Marioni).
\vspace*{-12pt}

%\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
\bibliography{ref}

\end{document}
