%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Welcome to Overleaf --- just edit your article on the left,
% and we'll compile it for you on the right. If you give
% someone the link to this page, they can edit at the same
% time. See the help menu above for more info. Enjoy!
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% For more detailed article preparation guidelines, please see:
% http://f1000research.com/author-guidelines

\documentclass[10pt,a4paper,twocolumn]{article}
\usepackage{f1000_styles}

%% Default: numerical citations
\usepackage[numbers]{natbib}

%% Uncomment this lines for superscript citations instead
% \usepackage[super]{natbib}

%% Uncomment these lines for author-year citations instead
% \usepackage[round]{natbib}
% \let\cite\citep

\begin{document}

\title{iSEE: Interactive SummarizedExperiment Explorer}
\titlenote{The title should be detailed enough for someone to know whether the article would be of interest to them, but also concise. Please ensure the broadness and claims within the title are appropriate to the content of the article itself.}
\author[1,$\dagger$]{Kevin Rue-Albrecht}
\author[2,3,$\dagger$]{Federico Marini}
\author[4,$\dagger$]{Charlotte Soneson}
\author[5,$\dagger$,$\ast$]{Aaron T. L. Lun}

\affil[1]{Kennedy Institute of Rheumatology, University of Oxford, Oxford OX3 7FY, United Kingdom}
\affil[2]{Center for Thrombosis and Hemostasis (CTH), University Medical Center of the Johannes Gutenberg University Mainz, Mainz}
\affil[3]{Institute for Medical Biostatistics, Epidemiology and Informatics (IMBEI), University Medical Center of the Johannes Gutenberg University Mainz, Mainz}
\affil[4]{University of Zurich and SIB Swiss Institute of Bioinformatics, Zurich, Switzerland}
\affil[5]{Cancer Research UK Cambridge Institute, University of Cambridge, Cambridge CB2 0RE, United Kingdom}
\affil[$\dagger$]{These authors contributed equally.}
\affil[$\ast$]{To whom correspondence should be addressed.}


\maketitle
\thispagestyle{fancy}

% Please list all authors that played a significant role in the research involved in the article. Please provide full affiliation information (including full institutional address, ZIP code and e-mail address) for all authors, and identify who is/are the corresponding author(s).
% TODO: full address and emails too? if so, in which format do we provide them?

\begin{abstract}

\textbf{Summary:} Data exploration is critical to the comprehension of large biological datasets generated by high-throughput assays such as sequencing.
However, most existing tools for interactive visualization are limited to specific assays or analyses.
Here, we present the iSEE (Interactive SummarizedExperiment Explorer) software package, which provides a general visual interface for exploring data in a SummarizedExperiment object.
iSEE is directly compatible with many existing R/Bioconductor packages for analyzing high-throughput biological data,
and provides useful features such as simultaneous examination of multiple data aspects, dynamic linking between plots and code tracking for reproducibility.
We demonstrate the utility and flexibility of iSEE by applying it to explore a range of real transcriptomics and proteomics datasets.\\
\textbf{Availability:} iSEE is publicly available as an R package from the open-source Bioconductor project (https://bioconductor.org/packages/iSEE).\\
\textbf{Contact:} aaron.lun@cruk.cam.ac.uk\\
\textbf{Supplementary information:} Supplementary data are available online.

% Abstracts should be up to 300 words and provide a succinct summary of the article. Although the abstract should explain why the article might be interesting, care should be taken not to inappropriately over-emphasize the importance of the work described in the article. Citations should not be used in the abstract, and the use of abbreviations should be minimized. If you are writing a Research or Systematic Review article, please structure your abstract into Background, Methods, Results, and Conclusions.


\end{abstract}

\section*{Keywords}

% Please list up to eight keywords to help readers interested in your article find it more easily.

Single-cell, gene expression, transcriptomics, proteomics, visualization, interactive, shiny, Bioconductor

% TODO: refine keywords

\clearpage

\section*{Introduction}
Interactive data exploration is critical to the analysis and comprehension of the data generated by high-throughput biological assays such as transcriptomics.
Exploration drives the formation of novel data-driven hypotheses prior to a more rigorous statistical analysis, and enables diagnosis of potential problems such as batch effects and low-quality samples.
To this end, visualization of the data in an intuitive and interactive interface is crucial, allowing researchers to examine the data from different perspectives across samples (e.g., experimental replicates, patients, single cells) and features (e.g., genes, transcripts, proteins, genomic regions, microarray probe sets).

Most existing tools for interactive visualization of biological data are designed for specific assays and analyses, e.g., pRoloc for proteomics \citep{gatto2014mass}, shinyMethyl for methylation \citep{fortin2014shinymethyl}, HTSvis for high-throughput screens \citep{scheeder2017htsvis}.
Opportunities for customization are generally limited, making it difficult to re-use the same visualization software for new technologies or experimental designs where different aspects of the data are of interest.
Moreover, standalone tools such as the Loupe Cell Browser from 10X Genomics \citep{zheng2017massively} do not easily integrate into established analysis pipelines such as those based on the R statistical programming language \citep{rcore2008R}.
This complicates any coordinated use of these tools with a reproducible, transparent, and statistically rigorous analysis.

\begin{figure*}[t]
\includegraphics[width=\textwidth]{pics/Fig1.pdf}
\caption{iSEE uses a customisable multi-panel layout (A) that simultaneously displays one or more panels of various types, where each panel type visualizes a different aspect of the data.
New panels of any type can be added (i), and all panels can be removed, reordered or resized (ii).
Panel types are available visualize sample-based reduced dimensionality embeddings (iii), sample-level metadata (iv), and experimental observations across samples for each feature (v).
For these panel types, each sample is represented as a point, which can be coloured according to sample-level metadata or feature-level observations.
Depending on whether the variables of interest are categorical or numeric, panels will automatically switch between scatter plots, violin plots or ``rectangle plots'' (where each combination of categorical levels is represented by a rectangle with area proportional to the frequency of combination).
Other panel types include row statistics tables (vi), to facilitate searching across features and their metadata; heatmaps (vii), to visualize experimental observations for multiple features; and feature-level metadata plots.
(B) Information can be transmitted between panels according to a user-specified scheme.
Here, the selection of feature $X$ in the row statistics table determines the y-axis of the feature expression plot, and colours the samples in the reduced dimension plot by the expression of $X$.
Selection of points in the reduced dimension plot (dotted blue line) also determines the samples that are shown in the column data (i.e., sample metadata) plot;
further selection of points in the column data plot determines the samples that are shown in the heatmap.
}
\label{fig:iSEE}
\end{figure*}

\section*{iSEE supports multiple genomics data types}
Here, we present the iSEE software package for interactive data exploration.
iSEE is implemented in R using the shiny framework \citep{chang2017shiny} and exploits data architectures from the open-source Bioconductor project \citep{gentleman2004bioconductor}.
The iSEE interface is initialized with a single function call accepting a SummarizedExperiment object as input \citep{huber2015orchestrating}.
The SummarizedExperiment class stores one or more matrices of experimental observations as ``assays'', where columns and rows represent samples and features, respectively.
For instance, these may represent gene expression matrices of raw counts along with different normalization strategies.
In addition, per-feature or per-sample variables are stored in the ``rowData'' and ``colData'', respectively; those may include experimental metadata as well as any analysis results.
All of these data aspects can be readily examined alongside one another in the multi-panel layout of the iSEE interface (Fig.~\ref{fig:iSEE}A).

Notably, the flexibility of the SummarizedExperiment class is the driving factor behind its broad deployment throughout the Bioconductor ecosystem, with uses in RNA sequencing \citep{love2014moderated}, methylation \citep{aryee2014minfi} and Hi-C analysis pipelines \citep{lun2016infrastructure}, amongst others.
As a consequence, the widespread adoption of the SummarizedExperiment class and its specialized extensions such as the SingleCellExperiment class enables iSEE to offer immediate interactive visualization for a variety of data modalities, complementing the state-of-the-art analysis methodologies already available in R/Bioconductor packages.
%In particular, we leverage off the state-of-the-art allowing us to avoid re-implementing them in iSEE (as is commonly done in other standalone visualization tools).
% KRA: any other child class like SingleCellExperiment worth mentioning? e.g. the BSseq class

\section*{Description of the iSEE interface}

In this section, give a brief description of each panel type.

% the previous section mainly focuses on the input object

\section*{Dynamic linking between panels}

A key feature of iSEE is the ability to dynamically transmit information between panels (Fig.~\ref{fig:iSEE}B).
Users can define arbitrary links between ``transmitting'' and ``receiving'' panels, whereby selections in transmitting tables and plots highlight the corresponding data points in receiving panels.
This feature facilitates exploration of the relationships between different aspects of the data -- for example, users can easily determine co-expression patterns of genes in a particular region of a reduced dimensionality embedding, by transmitting information from the reduced dimension plot to one or more feature expression plots.
The linking paradigm extends to multiple plots whereby a plot can transmit to multiple receivers, or a receiving plot can itself transmit its own selection to another plot.
This allows users to mimic the arbitrarily complex gating strategies often found in analyses of flow cytometry data \citep{finak2014opencyto}, with the advantage of being extensible to any (meta)data present in a SummarizedExperiment object.
Information can similarly be transferred between tables and plots, e.g., selecting a row determines how points are coloured in receiving plots.

\section*{Advanced features and customization}

% KRA: Also, we should probably highlight somewhere that additional inputs can customize the startup state of the application

\subsection*{ExperimentColorMap: custom color maps}

iSEE also offers a host of other customizable features to improve data exploration.
Users can supply their own colour mappings to iSEE, to control how colour scales are chosen for each metadata variable and assay type across all panels in the application.
% TODO: KRA develop here

\subsection*{Linked feature annotations}

Row statistics tables can be augmented with dynamic annotation based on the selected row, linking to online resources such as Ensembl or Entrez. % do we need to cite them? at least ENSEMBL has updated citable papers

\subsection*{Density-dependent downsampling for speed}

For large datasets, points can be downsampled in a density-dependent manner to accelerate rendering of the plots, improving the responsiveness of the interface without compromising the fidelity of the visualization.

\subsection*{Code tracking and reproducibility}

After analyses are completed, users can deploy iSEE on a server with a bespoke step-by-step ``tour'' of their dataset, guiding the audience through an examination of the salient features in the data.
iSEE also automatically memorizes the exact R code that was used to generate every plot and the existing links,
extending previous work by \cite{marini2016interrepro}.
This allows users to easily reproduce the results of any exploratory analysis, simply by copying the code reported by iSEE into their own analysis scripts.

\section*{Demonstration data sets and tours}

\subsection*{Single-cell plate-based RNA-seq}
% TODO: better title, or do we drop this allen section altogether?

We demonstrate the application of iSEE to examine a subset of 379 cells from of mouse visual cortex \citep{tasic2016allen}.

% TODO: embed the allen demo data set

\subsection*{Single-cell droplet-based RNA-seq}
% TODO: better title, or do we drop the allen section altogether (above)?

We also demonstrate the use of iSEE to explore a larger data set of 4,000 PBMC single-cell RNA sequencing dataset from 10X Genomics \citep{zheng2017massively}.

% embed the allen demo data set

\subsection*{Bulk RNA-seq}

We also demonstrate how iSEE seamlessly scales up with larger data sets as the TCGA RNA-seq collection of measurements for 23,368 genes across 7,706 tumor samples \citep{piccolo2015TCGA}.

% TODO: embed the TCGA demo data set

\subsection*{CyTOF}

Lastly, we also showcase the use of iSEE to vizualise and refine gating a gating analysis, followed by the investigation of a functional marker between two sample groups, using a published mass cytometry (CyTOF) data set \citep{bodenmiller2012cytof}.

% TODO: embed the TCGA demo data set

\section*{Conclusion}
iSEE provides a general interactive interface for visual exploration of a diverse range of `-omics datasets.
We provide a number of demonstrations at https://github.com/LTLA/iSEE2018, using iSEE to explore the TCGA RNA-seq collection \citep{piccolo2015TCGA}, the 4,000 PBMC single-cell RNA sequencing dataset from 10X Genomics \citep{zheng2017massively} and a published mass cytometry dataset consisting of over 170,000 cells \citep{bodenmiller2012cytof}.
% TODO: repurpose the conclusion if we showcase the examples in earlier sections

\subsection*{Author contributions}
In order to give appropriate credit to each author of an article, the individual
contributions of each author to the manuscript should be detailed in this section. We
recommend using author initials and then stating briefly how they contributed.

\subsection*{Competing interests}
All financial, personal, or professional competing interests for any of the authors that
could be construed to unduly influence the content of the article must be disclosed and
will be displayed alongside the article.

\subsection*{Grant information}
ATLL was supported by core funding from Cancer Research UK (award no. 17197 to Dr.\ John Marioni).
The work of FM is supported by the German Federal Ministry of Education and Research (BMBF 01EO1003).

\subsection*{Acknowledgements}
We thank the organizers and participants of the European Bioconductor Meeting 2017, where the idea for this package was first conceived.
We also thank members of the Bioconductor community for their helpful suggestions.


{\small\bibliographystyle{unsrtnat}
\bibliography{ref}}

\bigskip
References can be listed in any standard referencing style that uses a numbering system
(i.e. not Harvard referencing style), and should be consistent between references within
a given article.

% For more author guidance please see:
% http://f1000research.com/author-guidelines


% When all authors are happy with the paper, use the
% ‘Submit to F1000Research' button from the menu above
% to submit directly to the open life science journal F1000Research.

% Please note that this template results in a draft pre-submission PDF document.
% Articles will be professionally typeset when accepted for publication.

% We hope you find the F1000Research Overleaf template useful,
% please let us know if you have any feedback using the help menu above.


\end{document}
